Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Joel Riou
+-/
+import Mathlib.Algebra.Category.ModuleCat.Presheaf
+import Mathlib.Algebra.Category.ModuleCat.Adjunctions
+
+/-! The free presheaf of modules on a presheaf of sets
+
+In this file, given a presheaf of rings `R` on a category `C`,
+we construct the functor
+`PresheafOfModules.free (Cᵒᵖ ⥤ Type u) ⥤ PresheafOfModules.{u} R`
+which sends a presheaf of types to the corresponding presheaf of free modules.
+`PresheafOfModules.freeAdjunction` shows that this functor is the left
+adjoint to the forget functor.
+
+## Notes
+
+This contribution was created as part of the AIM workshop
+"Formalizing algebraic geometry" in June 2024.
+
+-/
+
+universe u v₁ u₁
+
+open CategoryTheory
+
+namespace PresheafOfModules
+
+variable {C : Type u₁} [Category.{v₁} C] (R : Cᵒᵖ ⥤ RingCat.{u})
+
+variable {R} in
+/-- Given a presheaf of types `F : Cᵒᵖ ⥤ Type u`, this is the presheaf
+of modules over `R` which sends `X : Cᵒᵖ` to the free `R.obj X`-module on `F.obj X`. -/
+@[simps]
+noncomputable def freeObj (F : Cᵒᵖ ⥤ Type u) : PresheafOfModules.{u} R where
+  obj := fun X ↦ (ModuleCat.free (R.obj X)).obj (F.obj X)
+  map := fun {X Y} f ↦ ModuleCat.freeDesc
+      (fun x ↦ ModuleCat.freeMk (F.map f x))
+  map_id := by aesop
+
+/-- The free presheaf of modules functors `(Cᵒᵖ ⥤ Type u) ⥤ PresheafOfModules.{u} R`. -/
+@[simps]
+noncomputable def free : (Cᵒᵖ ⥤ Type u) ⥤ PresheafOfModules.{u} R where
+  obj := freeObj
+  map {F G} φ :=
+    { app := fun X ↦ (ModuleCat.free (R.obj X)).map (φ.app X)
+      naturality := fun {X Y} f ↦ by
+        dsimp
+        ext x
+        simp only [ModuleCat.coe_comp, Function.comp_apply, ModuleCat.freeDesc_apply,
+          ModuleCat.restrictScalars.map_apply, ModuleCat.free_map_apply]
+        congr 1
+        exact NatTrans.naturality_apply φ f x }
```

Comment:
```suggestion
        simp [FunctorToTypes.naturality] }
```

Reply:
Maybe `FunctorToTypes.naturality` could be `@[simp]`?

Reply:
I do not have a particular opinion about this attribute in this case.